File: src/backend/agents/scheduling-agent.ts
/**
 * Scheduling AI Agent
 * 
 * Backend service for self-optimizing batch scheduling including:
 * - Batch Grouping by drug and density compatibility
 * - Cleaning Optimization to minimize changeover time
 * - Resource Allocation and constraint checking
 * - Schedule Optimization with efficiency scoring
 * 
 * This module can be deployed as a standalone Node.js service
 * or integrated with cloud functions (AWS Lambda, Azure Functions, etc.)
 */

import type {
  BatchOrderInput,
  ScheduleGroupOutput,
  ProductionConditionInput,
  ScheduleOptimizationOutput,
  ResourceConstraintInput,
  ScheduleValidationOutput,
  SchedulingAgentService,
} from './types';

// ============================================================
// CONFIGURATION
// ============================================================

const CONFIG = {
  // Cleaning time settings (minutes)
  NO_CLEANING_TIME: 0,
  PARTIAL_CLEANING_TIME: 15,
  FULL_CLEANING_TIME: 45,
  
  // Time savings per optimization
  SAME_DRUG_SAME_DENSITY_SAVINGS: 15,  // minutes per batch
  SAME_DRUG_DIFF_DENSITY_SAVINGS: 8,
  DIFF_DRUG_DIFF_DENSITY_SAVINGS: 5,
  
  // Constraint thresholds
  MIN_OPERATOR_SKILL_LEVEL: 2,  // 1=junior, 2=senior, 3=specialist
  MAX_MACHINE_WEAR_PERCENT: 90,
  MIN_ROOM_CLEARANCE_HOURS: 2,
  
  // Efficiency targets
  TARGET_EFFICIENCY_GAIN: 40,  // %
  MIN_BATCH_GROUP_SIZE: 2,
};

// ============================================================
// BATCH GROUPING ENGINE
// ============================================================

/**
 * Groups batches by drug and density compatibility for optimal sequencing
 */
export function groupBatches(batches: BatchOrderInput[]): ScheduleGroupOutput[] {
  const groups: ScheduleGroupOutput[] = [];
  
  // Sort batches by drug and density for grouping
  const sortedBatches = [...batches].sort((a, b) => {
    if (a.drug !== b.drug) return a.drug.localeCompare(b.drug);
    return a.density.localeCompare(b.density);
  });

  // Group 1: Same Drug + Same Density (no cleaning required)
  const sameDrugSameDensity = groupByDrugAndDensity(sortedBatches, true, true);
  if (sameDrugSameDensity.length >= CONFIG.MIN_BATCH_GROUP_SIZE) {
    groups.push({
      id: generateId(),
      type: 'same-drug-same-density',
      label: 'Same Drug + Same Density',
      batches: sameDrugSameDensity,
      cleaningRequired: 'none',
      cleaningTimeMinutes: CONFIG.NO_CLEANING_TIME,
      estimatedSavings: sameDrugSameDensity.length * CONFIG.SAME_DRUG_SAME_DENSITY_SAVINGS,
      sequenceOrder: 1,
      color: 'emerald',
    });
  }

  // Group 2: Same Drug + Different Density (partial cleaning)
  const sameDrugDiffDensity = groupByDrugAndDensity(sortedBatches, true, false);
  if (sameDrugDiffDensity.length >= CONFIG.MIN_BATCH_GROUP_SIZE) {
    groups.push({
      id: generateId(),
      type: 'same-drug-diff-density',
      label: 'Same Drug + Different Density',
      batches: sameDrugDiffDensity,
      cleaningRequired: 'partial',
      cleaningTimeMinutes: CONFIG.PARTIAL_CLEANING_TIME,
      estimatedSavings: sameDrugDiffDensity.length * CONFIG.SAME_DRUG_DIFF_DENSITY_SAVINGS,
      sequenceOrder: 2,
      color: 'amber',
    });
  }

  // Group 3: Different Drug + Different Density (full cleaning + QA clearance)
  const diffDrugDiffDensity = groupByDrugAndDensity(sortedBatches, false, false);
  if (diffDrugDiffDensity.length > 0) {
    groups.push({
      id: generateId(),
      type: 'diff-drug-diff-density',
      label: 'Different Drug + Different Density',
      batches: diffDrugDiffDensity,
      cleaningRequired: 'full',
      cleaningTimeMinutes: CONFIG.FULL_CLEANING_TIME,
      estimatedSavings: diffDrugDiffDensity.length * CONFIG.DIFF_DRUG_DIFF_DENSITY_SAVINGS,
      sequenceOrder: 3,
      color: 'rose',
    });
  }

  return groups;
}

/**
 * Helper function to group batches by drug and density criteria
 */
function groupByDrugAndDensity(
  batches: BatchOrderInput[],
  sameDrug: boolean,
  sameDensity: boolean
): BatchOrderInput[] {
  if (batches.length === 0) return [];

  const result: BatchOrderInput[] = [];
  const used = new Set<string>();

  for (let i = 0; i < batches.length; i++) {
    if (used.has(batches[i].id)) continue;

    const group: BatchOrderInput[] = [batches[i]];
    used.add(batches[i].id);

    for (let j = i + 1; j < batches.length; j++) {
      if (used.has(batches[j].id)) continue;

      const drugMatch = batches[i].drug === batches[j].drug;
      const densityMatch = batches[i].density === batches[j].density;

      if (sameDrug && sameDensity && drugMatch && densityMatch) {
        group.push(batches[j]);
        used.add(batches[j].id);
      } else if (sameDrug && !sameDensity && drugMatch && !densityMatch) {
        group.push(batches[j]);
        used.add(batches[j].id);
      } else if (!sameDrug && !sameDensity && !drugMatch) {
        group.push(batches[j]);
        used.add(batches[j].id);
      }
    }

    if (group.length >= 1) {
      result.push(...group);
    }
  }

  return result;
}

// ============================================================
// SCHEDULE OPTIMIZATION ENGINE
// ============================================================

/**
 * Optimizes the production schedule based on grouped batches and constraints
 */
export function optimizeSchedule(
  groups: ScheduleGroupOutput[],
  conditions: ProductionConditionInput[],
  constraints: ResourceConstraintInput = {}
): ScheduleOptimizationOutput {
  // Calculate baseline (unoptimized) time
  const totalBatches = groups.reduce((sum, g) => sum + g.batches.length, 0);
  const baselineCleaningTime = totalBatches * CONFIG.FULL_CLEANING_TIME;
  
  // Calculate optimized cleaning time
  const optimizedCleaningTime = groups.reduce((sum, g) => {
    return sum + (g.batches.length * g.cleaningTimeMinutes);
  }, 0);

  // Calculate total savings
  const totalSavingsMinutes = groups.reduce((sum, g) => sum + g.estimatedSavings, 0);
  const efficiencyGain = baselineCleaningTime > 0 
    ? Math.round((1 - optimizedCleaningTime / baselineCleaningTime) * 100)
    : 0;

  // Check production conditions for blockers
  const blockers = conditions.filter(c => c.status === 'blocked');
  const warnings = conditions.filter(c => c.status === 'warning');

  // Validate resource constraints
  const constraintViolations = validateConstraints(constraints, conditions);

  // Calculate schedule confidence
  const confidence = calculateScheduleConfidence(blockers, warnings, constraintViolations);

  // Generate optimization insights
  const insights = generateOptimizationInsights(groups, efficiencyGain, blockers);

  return {
    groups,
    totalBatches,
    totalSavingsMinutes,
    efficiencyGain,
    baselineCleaningTime,
    optimizedCleaningTime,
    blockers: blockers.map(b => ({
      unit: b.unit,
      name: b.name,
      detail: b.detail,
    })),
    warnings: warnings.map(w => ({
      unit: w.unit,
      name: w.name,
      detail: w.detail,
    })),
    constraintViolations,
    confidence,
    insights,
    isOptimal: blockers.length === 0 && constraintViolations.length === 0,
  };
}

/**
 * Validates resource constraints against current conditions
 */
function validateConstraints(
  constraints: ResourceConstraintInput,
  conditions: ProductionConditionInput[]
): string[] {
  const violations: string[] = [];

  // Check operator skill constraints
  if (constraints.minOperatorSkill && constraints.minOperatorSkill > CONFIG.MIN_OPERATOR_SKILL_LEVEL) {
    const operatorCondition = conditions.find(c => c.name.includes('Operator'));
    if (operatorCondition && operatorCondition.status !== 'ready') {
      violations.push('Insufficient operator skill level available');
    }
  }

  // Check machine wear constraints
  if (constraints.maxMachineWear) {
    const machineWearCondition = conditions.find(c => c.name.includes('Machine Wear'));
    if (machineWearCondition && machineWearCondition.status === 'warning') {
      violations.push('Machine wear approaching maximum threshold');
    }
  }

  // Check room clearance
  const roomCondition = conditions.find(c => c.name.includes('Room Clearance'));
  if (roomCondition && roomCondition.status !== 'ready') {
    violations.push('Room clearance not yet approved');
  }

  return violations;
}

/**
 * Calculates schedule confidence based on blockers and warnings
 */
function calculateScheduleConfidence(
  blockers: ProductionConditionInput[],
  warnings: ProductionConditionInput[],
  violations: string[]
): number {
  let confidence = 1.0;
  
  // Reduce confidence for blockers
  confidence -= blockers.length * 0.25;
  
  // Reduce confidence for warnings
  confidence -= warnings.length * 0.10;
  
  // Reduce confidence for constraint violations
  confidence -= violations.length * 0.15;

  return Math.max(0, Math.min(1, Math.round(confidence * 100) / 100));
}

/**
 * Generates optimization insights for the schedule
 */
function generateOptimizationInsights(
  groups: ScheduleGroupOutput[],
  efficiencyGain: number,
  blockers: ProductionConditionInput[]
): string[] {
  const insights: string[] = [];

  // Efficiency insight
  if (efficiencyGain >= CONFIG.TARGET_EFFICIENCY_GAIN) {
    insights.push(`Optimal grouping achieved ${efficiencyGain}% cleaning time reduction`);
  } else if (efficiencyGain > 0) {
    insights.push(`Current grouping provides ${efficiencyGain}% efficiency gain. Consider batch resequencing for further improvement.`);
  }

  // Group-specific insights
  const sameDrugGroup = groups.find(g => g.type === 'same-drug-same-density');
  if (sameDrugGroup && sameDrugGroup.batches.length >= 5) {
    insights.push(`Large same-product run (${sameDrugGroup.batches.length} batches) maximizes throughput`);
  }

  // Blocker insights
  if (blockers.length > 0) {
    insights.push(`Schedule may be delayed due to ${blockers.length} blocking condition(s)`);
  }

  // Cleaning optimization insight
  const noCleanGroups = groups.filter(g => g.cleaningRequired === 'none');
  if (noCleanGroups.length > 0) {
    const batchCount = noCleanGroups.reduce((sum, g) => sum + g.batches.length, 0);
    insights.push(`${batchCount} batches require no cleaning changeover`);
  }

  return insights;
}

// ============================================================
// SCHEDULE VALIDATION ENGINE
// ============================================================

/**
 * Validates a proposed schedule against all constraints and conditions
 */
export function validateSchedule(
  groups: ScheduleGroupOutput[],
  conditions: ProductionConditionInput[],
  equipmentFailures: { lineId: string; processName: string }[] = []
): ScheduleValidationOutput {
  const issues: Array<{ severity: 'warning' | 'error'; message: string }> = [];
  
  // Check for equipment failures
  for (const failure of equipmentFailures) {
    issues.push({
      severity: 'error',
      message: `${failure.processName} on ${failure.lineId} is offline - batches diverted to backup`,
    });
  }

  // Check production conditions
  for (const condition of conditions) {
    if (condition.status === 'blocked') {
      issues.push({
        severity: 'error',
        message: `${condition.name} at ${condition.unit}: ${condition.detail}`,
      });
    } else if (condition.status === 'warning') {
      issues.push({
        severity: 'warning',
        message: `${condition.name} at ${condition.unit}: ${condition.detail}`,
      });
    }
  }

  // Validate batch sequence
  for (const group of groups) {
    if (group.batches.length === 0) {
      issues.push({
        severity: 'warning',
        message: `Empty batch group: ${group.label}`,
      });
    }
  }

  // Check for QA clearance on full cleaning groups
  const fullCleanGroups = groups.filter(g => g.cleaningRequired === 'full');
  for (const group of fullCleanGroups) {
    const qaCondition = conditions.find(c => 
      c.name.includes('Room Clearance') || c.name.includes('QA')
    );
    if (qaCondition && qaCondition.status !== 'ready') {
      issues.push({
        severity: 'warning',
        message: `QA clearance pending for ${group.label} group`,
      });
    }
  }

  const errorCount = issues.filter(i => i.severity === 'error').length;
  const warningCount = issues.filter(i => i.severity === 'warning').length;

  return {
    isValid: errorCount === 0,
    canProceed: errorCount === 0 || equipmentFailures.length > 0, // Can proceed with backup
    issues,
    errorCount,
    warningCount,
    recommendations: generateValidationRecommendations(issues, equipmentFailures),
  };
}

/**
 * Generates recommendations based on validation issues
 */
function generateValidationRecommendations(
  issues: Array<{ severity: 'warning' | 'error'; message: string }>,
  equipmentFailures: { lineId: string; processName: string }[]
): string[] {
  const recommendations: string[] = [];

  if (equipmentFailures.length > 0) {
    recommendations.push('Continue production using backup equipment while maintenance addresses failures');
    recommendations.push('Monitor backup equipment capacity to prevent overload');
  }

  const errorIssues = issues.filter(i => i.severity === 'error');
  if (errorIssues.length > 0 && equipmentFailures.length === 0) {
    recommendations.push('Resolve blocking conditions before proceeding with scheduled batches');
  }

  const warningIssues = issues.filter(i => i.severity === 'warning');
  if (warningIssues.length > 0) {
    recommendations.push('Review and address warnings to maintain optimal schedule efficiency');
  }

  if (issues.length === 0) {
    recommendations.push('All conditions nominal - proceed with optimized schedule');
  }

  return recommendations;
}

// ============================================================
// UTILITY FUNCTIONS
// ============================================================

function generateId(): string {
  return Math.random().toString(36).substr(2, 9);
}

// ============================================================
// SERVICE EXPORT
// ============================================================

/**
 * Scheduling Agent Service
 * Complete service implementation that can be used as a standalone module
 */
export const SchedulingAgent: SchedulingAgentService = {
  groupBatches,
  optimizeSchedule,
  validateSchedule,
};

export default SchedulingAgent;

File: src/backend-python/agents/scheduling_agent.py
 """
 Scheduling AI Agent
 
 Backend service for self-optimizing batch scheduling including:
 - Batch Grouping by drug and density compatibility
 - Cleaning Optimization to minimize changeover time
 - Resource Allocation and constraint checking
 - Schedule Optimization with efficiency scoring
 """
 import uuid
 from typing import List, Optional
 from models.scheduling import (
     BatchOrderInput,
     ScheduleGroupOutput,
     ProductionConditionInput,
     ScheduleOptimizationOutput,
     ResourceConstraintInput,
     ScheduleValidationOutput,
 )
 
 
 class Config:
     NO_CLEANING_TIME = 0
     PARTIAL_CLEANING_TIME = 15
     FULL_CLEANING_TIME = 45
     SAME_DRUG_SAME_DENSITY_SAVINGS = 15
     SAME_DRUG_DIFF_DENSITY_SAVINGS = 8
     DIFF_DRUG_DIFF_DENSITY_SAVINGS = 5
     MIN_OPERATOR_SKILL_LEVEL = 2
     MAX_MACHINE_WEAR_PERCENT = 90
     MIN_ROOM_CLEARANCE_HOURS = 2
     TARGET_EFFICIENCY_GAIN = 40
     MIN_BATCH_GROUP_SIZE = 2
 
 
 def generate_id() -> str:
     return str(uuid.uuid4())[:9]
 
 
 def group_by_drug_and_density(
     batches: List[BatchOrderInput],
     same_drug: bool,
     same_density: bool,
 ) -> List[BatchOrderInput]:
     """Helper function to group batches by drug and density criteria"""
     if not batches:
         return []
 
     result = []
     used = set()
 
     for i, batch_i in enumerate(batches):
         if batch_i.id in used:
             continue
 
         group = [batch_i]
         used.add(batch_i.id)
 
         for j in range(i + 1, len(batches)):
             batch_j = batches[j]
             if batch_j.id in used:
                 continue
 
             drug_match = batch_i.drug == batch_j.drug
             density_match = batch_i.density == batch_j.density
 
             if same_drug and same_density and drug_match and density_match:
                 group.append(batch_j)
                 used.add(batch_j.id)
             elif same_drug and not same_density and drug_match and not density_match:
                 group.append(batch_j)
                 used.add(batch_j.id)
             elif not same_drug and not same_density and not drug_match:
                 group.append(batch_j)
                 used.add(batch_j.id)
 
         if len(group) >= 1:
             result.extend(group)
 
     return result
 
 
 def validate_constraints(
     constraints: ResourceConstraintInput,
     conditions: List[ProductionConditionInput],
 ) -> List[str]:
     """Validates resource constraints against current conditions"""
     violations = []
 
     if constraints.min_operator_skill and constraints.min_operator_skill > Config.MIN_OPERATOR_SKILL_LEVEL:
         operator_condition = next((c for c in conditions if "Operator" in c.name), None)
         if operator_condition and operator_condition.status != "ready":
             violations.append("Insufficient operator skill level available")
 
     if constraints.max_machine_wear:
         machine_wear_condition = next((c for c in conditions if "Machine Wear" in c.name), None)
         if machine_wear_condition and machine_wear_condition.status == "warning":
             violations.append("Machine wear approaching maximum threshold")
 
     room_condition = next((c for c in conditions if "Room Clearance" in c.name), None)
     if room_condition and room_condition.status != "ready":
         violations.append("Room clearance not yet approved")
 
     return violations
 
 
 def calculate_schedule_confidence(
     blockers: List[ProductionConditionInput],
     warnings: List[ProductionConditionInput],
     violations: List[str],
 ) -> float:
     """Calculates schedule confidence based on blockers and warnings"""
     confidence = 1.0
     confidence -= len(blockers) * 0.25
     confidence -= len(warnings) * 0.10
     confidence -= len(violations) * 0.15
     return max(0, min(1, round(confidence, 2)))
 
 
 def generate_optimization_insights(
     groups: List[ScheduleGroupOutput],
     efficiency_gain: float,
     blockers: List[ProductionConditionInput],
 ) -> List[str]:
     """Generates optimization insights for the schedule"""
     insights = []
 
     if efficiency_gain >= Config.TARGET_EFFICIENCY_GAIN:
         insights.append(f"Optimal grouping achieved {efficiency_gain}% cleaning time reduction")
     elif efficiency_gain > 0:
         insights.append(f"Current grouping provides {efficiency_gain}% efficiency gain. Consider batch resequencing for further improvement.")
 
     same_drug_group = next((g for g in groups if g.type == "same-drug-same-density"), None)
     if same_drug_group and len(same_drug_group.batches) >= 5:
         insights.append(f"Large same-product run ({len(same_drug_group.batches)} batches) maximizes throughput")
 
     if blockers:
         insights.append(f"Schedule may be delayed due to {len(blockers)} blocking condition(s)")
 
     no_clean_groups = [g for g in groups if g.cleaning_required == "none"]
     if no_clean_groups:
         batch_count = sum(len(g.batches) for g in no_clean_groups)
         insights.append(f"{batch_count} batches require no cleaning changeover")
 
     return insights
 
 
 def generate_validation_recommendations(
     issues: List[dict],
     equipment_failures: List[dict],
 ) -> List[str]:
     """Generates recommendations based on validation issues"""
     recommendations = []
 
     if equipment_failures:
         recommendations.append("Continue production using backup equipment while maintenance addresses failures")
         recommendations.append("Monitor backup equipment capacity to prevent overload")
 
     error_issues = [i for i in issues if i["severity"] == "error"]
     if error_issues and not equipment_failures:
         recommendations.append("Resolve blocking conditions before proceeding with scheduled batches")
 
     warning_issues = [i for i in issues if i["severity"] == "warning"]
     if warning_issues:
         recommendations.append("Review and address warnings to maintain optimal schedule efficiency")
 
     if not issues:
         recommendations.append("All conditions nominal - proceed with optimized schedule")
 
     return recommendations
 
 
 class SchedulingAgent:
     """Scheduling Agent Service for self-optimizing batch scheduling"""
 
     @staticmethod
     def group_batches(batches: List[BatchOrderInput]) -> List[ScheduleGroupOutput]:
         """Groups batches by drug and density compatibility for optimal sequencing"""
         groups = []
         
         sorted_batches = sorted(batches, key=lambda b: (b.drug, b.density))
 
         # Group 1: Same Drug + Same Density
         same_drug_same_density = group_by_drug_and_density(sorted_batches, True, True)
         if len(same_drug_same_density) >= Config.MIN_BATCH_GROUP_SIZE:
             groups.append(ScheduleGroupOutput(
                 id=generate_id(),
                 type="same-drug-same-density",
                 label="Same Drug + Same Density",
                 batches=same_drug_same_density,
                 cleaning_required="none",
                 cleaning_time_minutes=Config.NO_CLEANING_TIME,
                 estimated_savings=len(same_drug_same_density) * Config.SAME_DRUG_SAME_DENSITY_SAVINGS,
                 sequence_order=1,
                 color="emerald",
             ))
 
         # Group 2: Same Drug + Different Density
         same_drug_diff_density = group_by_drug_and_density(sorted_batches, True, False)
         if len(same_drug_diff_density) >= Config.MIN_BATCH_GROUP_SIZE:
             groups.append(ScheduleGroupOutput(
                 id=generate_id(),
                 type="same-drug-diff-density",
                 label="Same Drug + Different Density",
                 batches=same_drug_diff_density,
                 cleaning_required="partial",
                 cleaning_time_minutes=Config.PARTIAL_CLEANING_TIME,
                 estimated_savings=len(same_drug_diff_density) * Config.SAME_DRUG_DIFF_DENSITY_SAVINGS,
                 sequence_order=2,
                 color="amber",
             ))
 
         # Group 3: Different Drug + Different Density
         diff_drug_diff_density = group_by_drug_and_density(sorted_batches, False, False)
         if diff_drug_diff_density:
             groups.append(ScheduleGroupOutput(
                 id=generate_id(),
                 type="diff-drug-diff-density",
                 label="Different Drug + Different Density",
                 batches=diff_drug_diff_density,
                 cleaning_required="full",
                 cleaning_time_minutes=Config.FULL_CLEANING_TIME,
                 estimated_savings=len(diff_drug_diff_density) * Config.DIFF_DRUG_DIFF_DENSITY_SAVINGS,
                 sequence_order=3,
                 color="rose",
             ))
 
         return groups
 
     @staticmethod
     def optimize_schedule(
         groups: List[ScheduleGroupOutput],
         conditions: List[ProductionConditionInput],
         constraints: Optional[ResourceConstraintInput] = None,
     ) -> ScheduleOptimizationOutput:
         """Optimizes the production schedule based on grouped batches and constraints"""
         constraints = constraints or ResourceConstraintInput()
         
         total_batches = sum(len(g.batches) for g in groups)
         baseline_cleaning_time = total_batches * Config.FULL_CLEANING_TIME
         
         optimized_cleaning_time = sum(
             len(g.batches) * g.cleaning_time_minutes for g in groups
         )
 
         total_savings_minutes = sum(g.estimated_savings for g in groups)
         efficiency_gain = (
             round((1 - optimized_cleaning_time / baseline_cleaning_time) * 100)
             if baseline_cleaning_time > 0 else 0
         )
 
         blockers = [c for c in conditions if c.status == "blocked"]
         warnings = [c for c in conditions if c.status == "warning"]
         constraint_violations = validate_constraints(constraints, conditions)
         confidence = calculate_schedule_confidence(blockers, warnings, constraint_violations)
         insights = generate_optimization_insights(groups, efficiency_gain, blockers)
 
         return ScheduleOptimizationOutput(
             groups=groups,
             total_batches=total_batches,
             total_savings_minutes=total_savings_minutes,
             efficiency_gain=efficiency_gain,
             baseline_cleaning_time=baseline_cleaning_time,
             optimized_cleaning_time=optimized_cleaning_time,
             blockers=[{"unit": b.unit, "name": b.name, "detail": b.detail} for b in blockers],
             warnings=[{"unit": w.unit, "name": w.name, "detail": w.detail} for w in warnings],
             constraint_violations=constraint_violations,
             confidence=confidence,
             insights=insights,
             is_optimal=len(blockers) == 0 and len(constraint_violations) == 0,
         )
 
     @staticmethod
     def validate_schedule(
         groups: List[ScheduleGroupOutput],
         conditions: List[ProductionConditionInput],
         equipment_failures: Optional[List[dict]] = None,
     ) -> ScheduleValidationOutput:
         """Validates a proposed schedule against all constraints and conditions"""
         equipment_failures = equipment_failures or []
         issues = []
 
         for failure in equipment_failures:
             issues.append({
                 "severity": "error",
                 "message": f"{failure['processName']} on {failure['lineId']} is offline - batches diverted to backup",
             })
 
         for condition in conditions:
             if condition.status == "blocked":
                 issues.append({
                     "severity": "error",
                     "message": f"{condition.name} at {condition.unit}: {condition.detail}",
                 })
             elif condition.status == "warning":
                 issues.append({
                     "severity": "warning",
                     "message": f"{condition.name} at {condition.unit}: {condition.detail}",
                 })
 
         for group in groups:
             if not group.batches:
                 issues.append({
                     "severity": "warning",
                     "message": f"Empty batch group: {group.label}",
                 })
 
         full_clean_groups = [g for g in groups if g.cleaning_required == "full"]
         for group in full_clean_groups:
             qa_condition = next(
                 (c for c in conditions if "Room Clearance" in c.name or "QA" in c.name),
                 None
             )
             if qa_condition and qa_condition.status != "ready":
                 issues.append({
                     "severity": "warning",
                     "message": f"QA clearance pending for {group.label} group",
                 })
 
         error_count = sum(1 for i in issues if i["severity"] == "error")
         warning_count = sum(1 for i in issues if i["severity"] == "warning")
 
         return ScheduleValidationOutput(
             is_valid=error_count == 0,
             can_proceed=error_count == 0 or len(equipment_failures) > 0,
             issues=issues,
             error_count=error_count,
             warning_count=warning_count,
             recommendations=generate_validation_recommendations(issues, equipment_failures),
         )